sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/RecommendationAIImportUserEvents.java

```diff
@@ -0,0 +1,216 @@
+    return this.toBuilder().setProjectId(projectId).build();
+  }
+
+  public RecommendationAIImportUserEvents withCatalogName(String catalogName) {
+    return this.toBuilder().setCatalogName(catalogName).build();
+  }
+
+  public RecommendationAIImportUserEvents withEventStore(String eventStore) {
+    return this.toBuilder().setEventStore(eventStore).build();
+  }
+
+  public RecommendationAIImportUserEvents withBatchSize(Integer batchSize) {
+    return this.toBuilder().setBatchSize(batchSize).build();
+  }
+
+  /**
+   * The transform converts the contents of input PCollection into {@link UserEvent}s and then calls
+   * the Recommendation AI service to create the user event.
+   *
+   * @param input input PCollection
+   * @return PCollection after transformations
+   */
+  @Override
+  public PCollectionTuple expand(PCollection<KV<String, GenericJson>> input) {
+    return input
+        .apply(
+            "Batch Contents",
+            GroupIntoBatches.<String, GenericJson>ofSize(batchSize())
+                .withMaxBufferingDuration(maxBufferingDuration())
+                .withShardedKey())
+        .apply(
+            "Import CatalogItems",
+            ParDo.of(new ImportUserEvents(projectId(), catalogName(), eventStore()))
+                .withOutputTags(SUCCESS_TAG, TupleTagList.of(FAILURE_TAG)));
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    /** @param projectId ID of Google Cloud project to be used for creating user events. */
+    public abstract Builder setProjectId(@Nullable String projectId);
+
+    /** @param catalogName Name of the catalog where the user events will be created. */
+    public abstract Builder setCatalogName(@Nullable String catalogName);
+
+    /** @param eventStore Name of the event store where the user events will be created. */
+    public abstract Builder setEventStore(@Nullable String eventStore);
+
+    /**
+     * @param batchSize Amount of input elements to be sent to Recommendation AI service in one
+     *     request.
+     */
+    public abstract Builder setBatchSize(Integer batchSize);
+
+    /**
+     * @param maxBufferingDuration Time limit (in processing time) on how long an incomplete batch
+     *     of elements is allowed to be buffered.
+     */
+    public abstract Builder setMaxBufferingDuration(Duration maxBufferingDuration);
+
+    public abstract RecommendationAIImportUserEvents build();
+  }
+
+  static class ImportUserEvents
+      extends DoFn<KV<ShardedKey<String>, Iterable<GenericJson>>, UserEvent> {
+    private final String projectId;
+    private final String catalogName;
+    private final String eventStore;
+
+    /**
+     * @param projectId ID of GCP project to be used for creating user events.
+     * @param catalogName Catalog name for UserEvent creation.
+     * @param eventStore Event store name for UserEvent creation.
+     */
+    public ImportUserEvents(String projectId, String catalogName, String eventStore) {
```

Comment:
you can remove `public` from this constructor so it can only be used within the package